### PR TITLE
Add Figure.timestamp to plot the GMT timestamp logo

### PR DESCRIPTION
**Pull request by seisman**
**Description of proposed changes**

In this PR, I add the `Figure.timestamp` method to plot the GMT timestamp logo, just like what GMT's `-U` option does. The final goal is to remove the `timestamp` alias from all the plotting methods (c.f. https://github.com/GenericMappingTools/pygmt/pull/2135 and https://github.com/GenericMappingTools/pygmt/issues/924#issuecomment-780963272).

Here are some notes:

1. The full syntax of the `-U` option is: `-U[<label>][+c][+j<just>][+o<dx>[/<dy>]]`. `-U+c` can be used to append the raw GMT command (e.g., `gmt coast -R... -J...`) after the timestamp logo. It's not implemented because (1) It's rarely used. A figure is usually produced by multiple commands, so it makes no sense to show the command string of a single command; (2) It makes little sense to show the raw GMT command string for a figure produced by a Python script;  (3) It's also impossible to implement it in the `Figure.timestamp()` function.
2. The `Figure.timestamp` method is implemented by calling `gmt plot -T -U...`
3. There are four GMT defaults (`FONT_LOGO`, `MAP_LOGO`, `MAP_LOGO_POS`, and `FORMAT_TIME_STAMP`) that can affect the timestamp logo. 
   - `MAP_LOGO` controls if a timestamp logo should be plotted. It makes no sense here, so ignored.
   - `MAP_LOGO_POS` defaults to `BL/-54p/-54p`. It's equivalent to setting the `+j` and `+o` modifiers. So it's ignored.
   - `FONT_LOGO` can control the font ID of the text strings. It's implemented by passing `--FONT_LOGO=xxx` to the GMT C API.
   - `FORMAT_TIME_STAMP` can control the format of the timestamp. It's implemented by passing `--FORMAT_TIME_STAMP` to the GMT C API.
4. An upstream bug of the `+o` modifier for GMT <=6.4.0 (https://github.com/GenericMappingTools/gmt/issues/7107).
5. For GMT<=6.4.0, `FONT_LOGO` can only change the font ID, but since GMT 6.5.0, it can also change both the font ID and color (but not the font size). See https://github.com/GenericMappingTools/gmt/issues/7125
6. In GMT 6.5.0, a new modifier `+t` will be added. See https://github.com/GenericMappingTools/gmt/issues/7125.
7. It's difficult to test this function, because the image has the timestamp string, which changes every second. The workaround here is passing a constant format string to `FORMAT_TIME_STAMP` (e.g., `--FORMAT_TIME_STAMP="9999-99-99T99:99:99"`).

**Examples:**
```
>>> # Plot the GMT timestamp logo.
>>> import pygmt
>>> fig = pygmt.Figure()
>>> fig.timestamp()
>>> fig.show()

>>> # Plot the GMT timestamp logo with a custom label.
>>> fig = pygmt.Figure()
>>> fig.timestamp(label="Powered by PyGMT")
>>> fig.show()
```

**Documentation preview:** https://pygmt-dev--2208.org.readthedocs.build/en/2208/api/generated/pygmt.Figure.timestamp.html#pygmt.Figure.timestamp

**Reminders**

- [x] Run `make format` and `make check` to make sure the code follows the style guide.
- [x] Add tests for new features or tests that would have caught the bug that you're fixing.
- [x] Add new public functions/methods/classes to `doc/api/index.rst`.
- [x] Write detailed docstrings for all functions/methods.
- [ ] If wrapping a new module, open a 'Wrap new GMT module' issue and submit reasonably-sized PRs.
- [ ] If adding new functionality, add an example to docstrings or tutorials.

**Slash Commands**

You can write slash commands (`/command`) in the first line of a comment to perform
specific operations. Supported slash commands are:

- `/format`: automatically format and lint the code
- `/test-gmt-dev`: run full tests on the latest GMT development version

### CODE REVIEW

File: pygmt/src/timestamp.py
Comment (seisman) [inline pygmt/src/timestamp.py:22]:
```suggestion
            # Workaround for an upstream bug which will be fixed in GMT 6.5.0
            # See https://github.com/GenericMappingTools/gmt/issues/7107
            kwargs["U"] += f"+o{offsets}/{offsets}"
```

File: pygmt/src/timestamp.py
Comment (seisman) [inline pygmt/src/timestamp.py:57]: Please note that we can't use decorator `@kwargs_to_strings(offset="sequence")` here, because the decorator can only deal with keyword arguments and doesn't work for the default values. I.e., it doesn't work for `Figure.timestamp()` but works for `Figure.timestamp(offset=["-54p", "-54p"])`, although `offset` defaults to `["-54p", "-54p"]`.

**Comment by seisman** [comment]
Ping @GenericMappingTools/pygmt-maintainers for comments on the current implementation of the `Figure.timestamp()` function and ideas about how to test it.

**Comment by seisman** [comment]
The [libfaketime](https://github.com/wolfcw/libfaketime) package can be used to temporarily change the system date/time. It's also available on [conda-forge](https://anaconda.org/conda-forge/libfaketime), but it's a CLI program.

There are Python wrapper (https://github.com/simon-weber/python-libfaketime) and pytest plugin (https://github.com/touilleMan/pytest-libfaketime), but both are not actively maintained.

File: pygmt/src/timestamp.py
Comment (seisman) [inline pygmt/src/timestamp.py:99]: Two thoughts when implementing this function:

1. Add a new parameter (e.g., `confdict`) to the `build_arg_string` function, so that we can pass a dict of GMT defaults to the `call_module` function, e.g.,
   ```
   build_arg_string(kwdict, confdict={"FONT_LOGO": font, "FORMAT_TIME_STAMP": timefmt}
   ```

2. The four parameters `FONT_LOGO`, `MAP_LOGO`, `MAP_LOGO_POS`, and `FORMAT_TIME_STAMP` won't be used after the `Figure.timestamp` function is used, so they can be removed from the `pygmt.config` keyword list:
 https://github.com/GenericMappingTools/pygmt/blob/9253af320ffbe90cc675acba33695a0325bea600/pygmt/src/config.py#L26-L28

**Comment by weiji14** [comment]
> 7\. It's difficult to write tests for this module, because the image has the timestamp string, which changes every second. We can't test it unless we can find a way to set the current time to a fixed value.

Maybe use `@check_figures_equal` following https://github.com/GenericMappingTools/pygmt/blob/v0.8.0/doc/contributing.md#using-check_figures_equal, and output only the date and not the precise time down to the second?

File: pygmt/src/timestamp.py
Comment (weiji14) [reply pygmt/src/timestamp.py:99]: Hmm, point 1 seems ok, but I'm not sure about point 2. I'd prefer to keep the three parameters in the list, because even if you remove it, you're only removing the autocompletion, but users can still manually set those parameters.

File: pygmt/src/timestamp.py
Comment (seisman) [reply pygmt/src/timestamp.py]: Related issue report at https://github.com/GenericMappingTools/pygmt/issues/2361.

File: pygmt/src/timestamp.py
Comment (seisman) [reply pygmt/src/timestamp.py]: Point 1 was implemented in PR #2324

File: pygmt/src/timestamp.py
Comment (seisman) [inline pygmt/src/timestamp.py:30]: I'm still unsure what's the best way to document features that require new GMT versions.

File: pygmt/src/timestamp.py
Comment (yvonnefroehlich) [inline pygmt/src/timestamp.py:32]:
```suggestion
        The text string shown after the GMT time stamp logo.
```

File: pygmt/src/timestamp.py
Comment (yvonnefroehlich) [inline pygmt/src/timestamp.py:34]:
```suggestion
        Justification of the time stamp. *justification* is a two-character code
```

File: pygmt/src/timestamp.py
Comment (yvonnefroehlich) [inline pygmt/src/timestamp.py:40]:
```suggestion
        Offset the anchor point of the time stamp by *offset_x* and *offset_y*.
```

File: pygmt/src/timestamp.py
Comment (yvonnefroehlich) [inline pygmt/src/timestamp.py:44]:
```suggestion
        Font of the time stamp and the optional label.
```

File: pygmt/src/timestamp.py
Comment (yvonnefroehlich) [inline pygmt/src/timestamp.py:54]:
```suggestion
    >>> # Plot the GMT time stamp logo.
```

File: pygmt/src/timestamp.py
Comment (yvonnefroehlich) [inline pygmt/src/timestamp.py:61]:
```suggestion
    >>> # Plot the GMT time stamp logo with a custom label.
```

File: pygmt/src/timestamp.py
Comment (yvonnefroehlich) [reply pygmt/src/timestamp.py:30]: Hm. Is there a similar situation for the Julia wrapper, which can serve as an orientation?

File: pygmt/src/timestamp.py
Comment (seisman) [reply pygmt/src/timestamp.py:22]: Wikipedia says "timestamp" (https://en.wikipedia.org/wiki/Timestamp).

Comment (michaelgrund) [reply pygmt/src/timestamp.py:22]: I think both are possible, but would also go with `timestamp` as is.

Comment (yvonnefroehlich) [reply pygmt/src/timestamp.py:22]: I am also fine with "timestamp" (no space).

File: pygmt/src/timestamp.py
Comment (yvonnefroehlich) [inline pygmt/src/timestamp.py:48]:
```suggestion
        Format of the time information in the UNIX timestamp. This format is
```

File: pygmt/src/timestamp.py
Comment (yvonnefroehlich) [inline pygmt/src/timestamp.py:28]:
```suggestion
        logo. Set this parameter to replace the UNIX timestamp with a
```

File: pygmt/src/timestamp.py
Comment (yvonnefroehlich) [inline pygmt/src/timestamp.py:27]:
```suggestion
        If ``None``, the current UNIX timestamp is shown in the GMT timestamp
```

File: pygmt/src/timestamp.py
Comment (michaelgrund) [inline pygmt/src/timestamp.py:34]:
```suggestion
        Justification of the timestamp. The *justification* is a two-character code
```

File: pygmt/tests/test_timestamp.py
Comment (michaelgrund) [inline pygmt/tests/test_timestamp.py:15]:
```suggestion
    Fake datetime that will be passed to the "timefmt" parameter, so that the
```

**Comment by seisman** [comment]
Ping @GenericMappingTools/pygmt-maintainers for reviews.

File: pygmt/tests/test_timestamp.py
Comment (weiji14) [inline pygmt/tests/test_timestamp.py:50]: The output of the justification image doesn't look correct?



All of the timestamps are on the bottom left (BL). Shouldn't they be at the four different corners?

File: pygmt/src/timestamp.py
Comment (weiji14) [inline pygmt/src/timestamp.py:89]: Not that we have a GMT 6.4.1, but this is more precise given the second line.

```suggestion
        if Version(__gmt_version__) < Version("6.5.0"):
            raise GMTInvalidInput("The parameter 'text' requires GMT>=6.5.0.")
```

File: pygmt/src/timestamp.py
Comment (weiji14) [inline pygmt/src/timestamp.py:46]: Should a GMTInvalidInput error be raised when users pass `font` as a parameter with GMT 6.3.0? Or is the error clear enough from GMT?

Comment (seisman) [reply pygmt/src/timestamp.py:46]: The `font` parameter works for any GMT versions. 

For GMT>6.4.0, it can change both the **font ID** and **font color**, but for GMT<=6.4.0, the font color is always black so it can only change the font ID.

File: pygmt/tests/test_timestamp.py
Comment (seisman) [reply pygmt/tests/test_timestamp.py:50]: Unlike other embellishments like inset, legend or direction rose, the timestamp logo are always located at the lower left corner of the map. I believe this is GMT's design.

See https://docs.generic-mapping-tools.org/dev/gmt.conf.html#term-MAP_LOGO_POS

> Sets the justification and the position of the logo/timestamp box **relative to the current plot’s lower left corner** (i.e., map origin)

File: pygmt/src/timestamp.py
Comment (weiji14) [reply pygmt/src/timestamp.py:46]: Ah ok, maybe reword it into something like this?
```suggestion
        Font of the timestamp and the optional label. The parameter can't
        change the font color for GMT<=6.4.0, only the font label.
```

Comment (seisman) [reply pygmt/src/timestamp.py:46]: I'm afraid "font label" is incorrect here. It should be "font ID" or "font name".

File: pygmt/tests/test_timestamp.py
Comment (weiji14) [reply pygmt/tests/test_timestamp.py:50]: Interesting design choice :sweat_smile: Actually, I just realized I was thinking of `position` rather than `justification`. Maybe we should document the default `offset="-54p/-54p"` under the offset docstring?

File: pygmt/src/timestamp.py
Comment (weiji14) [reply pygmt/src/timestamp.py:46]: Owh, you mean like font style (e.g. Helvetica)? Feel free to edit the suggested changes.

File: pygmt/tests/test_timestamp.py
Comment (seisman) [reply pygmt/tests/test_timestamp.py:50]: The default values are clear from the function definition:

`Figure.timestamp((text=None, label=None, justification='BL', offset=('-54p', '-54p'), font='Helvetica,black', timefmt='%Y %b %d %H:%M:%S')`

I'm hesitate to document the defaults in the docstring.

Comment (weiji14) [reply pygmt/tests/test_timestamp.py:50]: Ah ok, was thinking of adding `[Default is ('-54p', '-54p')]`, but it's not really needed I guess.

**Review by weiji14 — APPROVED** [review]
Only a very minor suggestion as mentioned in https://github.com/GenericMappingTools/pygmt/pull/2208/files#r1122590265 (up to you to apply), but otherwise wait for second reviewer to approve.

File: pygmt/src/timestamp.py
Comment (seisman) [inline pygmt/src/timestamp.py:46]:
```suggestion
        change the font color for GMT<=6.4.0, only the font ID.
```

Comment (seisman) [reply pygmt/src/timestamp.py]: Added "only the font ID" in 007964375e091a93c8e7c63c840fb3affc2a8e01.

**Comment by michaelgrund** [comment]
Would you also like to add a gallery example or should we do this in a separate PR @seisman ?

**Comment by seisman** [comment]
> Would you also like to add a gallery example or should we do this in a separate PR @seisman ?

It's not on my todo list, so better to wait for someone else to work on the gallery example in another PR.

**Pull request merged**
